x/tx/aminojson/json_marshal.go

```diff
@@ -0,0 +1,357 @@
+			"empty_string":     emptyStringEncoder,
+			"json_default":     jsonDefaultEncoder,
+			"legacy_coins":     nullSliceAsEmptyEncoder,
+			"cosmos_dec_bytes": cosmosDecBytesEncoder,
+		},
+	}
+	return aj
+}
+
+// DefineMessageEncoding defines a custom encoding for a protobuf message.
+func (aj AminoJSON) DefineMessageEncoding(name string, encoder MessageEncoder) {
+	aj.messageEncoders[name] = encoder
+}
+
+// DefineFieldEncoding defines a custom encoding for a protobuf field.
+func (aj AminoJSON) DefineFieldEncoding(name string, encoder FieldEncoder) {
+	aj.fieldEncoders[name] = encoder
+}
+
+// MarshalAmino serializes a protobuf message to JSON.
+func (aj AminoJSON) MarshalAmino(message proto.Message) ([]byte, error) {
+	buf := &bytes.Buffer{}
+	err := aj.beginMarshal(message.ProtoReflect(), buf)
+	return buf.Bytes(), err
+}
+
+func (aj AminoJSON) beginMarshal(msg protoreflect.Message, writer io.Writer) error {
+	name, named := getMessageName(msg)
+	if named {
+		_, err := writer.Write([]byte(fmt.Sprintf(`{"type":"%s","value":`, name)))
+		if err != nil {
+			return err
+		}
+	}
+
+	err := aj.marshal(protoreflect.ValueOfMessage(msg), writer)
+	if err != nil {
+		return err
+	}
+
+	if named {
+		_, err = writer.Write([]byte("}"))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (aj AminoJSON) marshal(value protoreflect.Value, writer io.Writer) error {
+	switch val := value.Interface().(type) {
+	case protoreflect.Message:
+		err := aj.marshalMessage(val, writer)
+		return err
+
+	case protoreflect.Map:
+		return errors.New("maps are not supported")
+
+	case protoreflect.List:
+		if !val.IsValid() {
+			_, err := writer.Write([]byte("null"))
+			return err
+		}
+		return aj.marshalList(val, writer)
+
+	case string, bool, int32, uint32, protoreflect.EnumNumber:
+		return jsonMarshal(writer, val)
+
+	case uint64, int64:
+		_, err := fmt.Fprintf(writer, `"%d"`, val) // quoted
+		return err
+
+	case []byte:
+		_, err := fmt.Fprintf(writer, `"%s"`, base64.StdEncoding.EncodeToString(val))
+		return err
+
+	default:
+		return errors.Errorf("unknown type %T", val)
+	}
+}
+
+func (aj AminoJSON) marshalMessage(msg protoreflect.Message, writer io.Writer) error {
+	if msg == nil {
+		return errors.New("nil message")
+	}
+
+	switch msg.Descriptor().FullName() {
+	case timestampFullName:
+		// replicate https://github.com/tendermint/go-amino/blob/8e779b71f40d175cd1302d3cd41a75b005225a7a/json-encode.go#L45-L51
+		return marshalTimestamp(msg, writer)
+	case durationFullName:
+		return marshalDuration(msg, writer)
+	case anyFullName:
+		return aj.marshalAny(msg, writer)
+	}
+
+	if encoder := aj.getMessageEncoder(msg); encoder != nil {
+		err := encoder(msg, writer)
+		return err
+	}
+
+	_, err := writer.Write([]byte("{"))
+	if err != nil {
+		return err
+	}
+
+	fields := msg.Descriptor().Fields()
+	first := true
+	emptyOneOfWritten := map[string]bool{}
+	for i := 0; i < fields.Len(); i++ {
+		f := fields.Get(i)
+		v := msg.Get(f)
```

Comment:
Do you know how go-amino handles field ordering? It reads the fields sequentially using go `reflect`, right? Which means this is correct. (also assuming our pulsar generator keeps the same field ordering between proto types and go types)

Reply:
Right, go-amino uses the `reflect` package.

Reply:
Whether ordering is the same comes down to gogo vs pulsar code generation. IME (from the work in this PR) they are the same.